### PR TITLE
Scrapper, Ekaterina Khomutova - 22FPL2

### PR DESCRIPTION
**Pull request by EkaterinaKhomutova**

### CODE REVIEW

**Comment by demid5111** [comment]
nothing to review

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -26,11 +74,47 @@ def _extract_config_content(self) -> ConfigDTO:
...
+            headless_mode=file['headless_mode']
+        )
+        return config
+
 
     def _validate_config_content(self) -> None:
         """
         Ensure configuration parameters are not corrupt.
         """
+        config = json.load(open(self._path_to_config))
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:95]: use dto. otherwise, why do we need it

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -26,11 +74,47 @@ def _extract_config_content(self) -> ConfigDTO:
...
+        if not isinstance(config['seed_urls'], list) or \
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:96]: we do not work with any website, right? we need seed urls to be exactly of the website we target

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -26,11 +74,47 @@ def _extract_config_content(self) -> ConfigDTO:
...
+        if not isinstance(num, int):
+            raise IncorrectNumberOfArticlesError
+
+        if not isinstance(config['headers'], dict):
+            raise IncorrectHeadersError
+
+        if not isinstance(config['encoding'], str):
+            raise IncorrectEncodingError
+
+        if not isinstance(config['timeout'], int) or not (0 < config['timeout'] < 60):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:113]: open conditions. we should avoid making `not (CONDITION and CONDITION)`, it is hard to read and understand

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: lab_5_scrapper/scrapper.py (hunk 1 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:97]: How to use use it? In the documentation it is said:
> "This method should be called during Config class instance initialization step before Config._extract_config_content() method call to check config fields and make sure they are appropriate and can be used inside the program."
Actually, this way, I don't understand how to use dto, if we should to not to create it yet

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:115]: fixed, I guess

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:98]: Should I add more conditions for this? There were no such info in the documentation

File: requirements.txt [hunk 4]
```diff
@@ -1 +1,3 @@
-
+requests==2.31.0
+bs4==0.0.2
```
Comment (demid5111) [inline requirements.txt:2]: .

File: requirements.txt [hunk 5]
```diff
@@ -1 +1,3 @@
...
+BeautifulSoup4==4.12.3
```
Comment (demid5111) [inline requirements.txt:3]: newline

File: lab_5_scrapper/settings.json [hunk 6]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 4
```
Comment (demid5111) [inline lab_5_scrapper/settings.json:2]: too low..

File: lab_5_scrapper/scrapper_config.json [hunk 7]
```diff
@@ -1,9 +1,13 @@
 {
-    "seed_urls": [],
-    "headers": {},
-    "total_articles_to_find_and_parse": 0,
-    "encoding": "",
-    "timeout": 0,
+    "seed_urls": ["https://www.gazetavechorka.ru/news/proisshestviya"],
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:2]: will it be enough to collect 100 articles?

File: lab_5_scrapper/scrapper_config.json [hunk 8]
```diff
@@ -1,9 +1,13 @@
...
+    "headers": {"Accept-Language": "ru,en;q=0.9",
```
Comment (demid5111) [inline lab_5_scrapper/scrapper_config.json:3]: make proper indents and formatting. example: https://jsonformatter.curiousconcept.com/

File: lab_5_scrapper/scrapper.py [hunk 9]
```diff
@@ -204,6 +323,12 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    assets_path = pathlib.Path(base_path) / 'ASSETS_PATH'
+    if not assets_path.exists():
+        assets_path.mkdir(parents=True, exist_ok=True)
+    else:
+        for file in assets_path.iterdir():
+            file.unlink()
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:331]: use shutil

File: lab_5_scrapper/scrapper.py [hunk 10]
```diff
@@ -204,6 +323,12 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    assets_path = pathlib.Path(base_path) / 'ASSETS_PATH'
+    if not assets_path.exists():
+        assets_path.mkdir(parents=True, exist_ok=True)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:328]: if it does not exist, then why do we need flag `exist_ok=True`?

File: lab_5_scrapper/scrapper.py [hunk 11]
```diff
@@ -127,11 +232,24 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+                    self.urls.append(href)
+                    return href
 
     def find_articles(self) -> None:
         """
         Find articles.
         """
+        seed_urls = self.config.get_seed_urls()
+        for url in seed_urls:
+            response = make_request(url, self.config)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:249]: where are pauses between requests?

File: lab_5_scrapper/scrapper.py [hunk 12]
```diff
@@ -127,11 +232,24 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+            if response.status_code < 400:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:250]: use built-in properties

File: lab_5_scrapper/scrapper.py [hunk 13]
```diff
@@ -127,11 +232,24 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+                soup = BeautifulSoup(response.text, 'html.parser')
+                [self._extract_url(soup) for _ in range(10)]
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:252]: ??

File: lab_5_scrapper/scrapper.py [hunk 14]
```diff
@@ -127,11 +232,24 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
 
     def find_articles(self) -> None:
         """
         Find articles.
         """
+        seed_urls = self.config.get_seed_urls()
+        for url in seed_urls:
+            response = make_request(url, self.config)
+            if response.status_code < 400:
+                soup = BeautifulSoup(response.text, 'html.parser')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:251]: how do you control number of articles that you collect and stop if they are collected

File: lab_5_scrapper/scrapper.py [hunk 15]
```diff
@@ -127,11 +232,24 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
         Returns:
             str: Url from HTML
         """
+        container_div = article_bs.find('div', class_='col-lg-9 col-md-12')
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:235]: is there a better strategy to find it?

File: lab_5_scrapper/scrapper.py [hunk 16]
```diff
@@ -127,11 +232,24 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
+        for link in container_div.find_all('a'):
+            if link.get('href').startswith('/news/202'):
+                href = 'https://www.gazetavechorka.ru' + link.get('href')
+                if href and href not in self.urls:
+                    self.urls.append(href)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:240]: extract URL does not change the state of the instance. only does simple thing: takes/constructs URL based on page content.

File: lab_5_scrapper/scrapper.py [hunk 17]
```diff
@@ -26,11 +84,43 @@ def _extract_config_content(self) -> ConfigDTO:
         Returns:
             ConfigDTO: Config values
         """
+        with open(self.path_to_config) as file:
+            file = json.load(file)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:88]: do not overwrite same variable. name it differently

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper.py:98]: it is logic. you can't work with any site. you specialize on it, therefore, all other websites should be rejected - a.k.a raised Exception

**Review by demid5111 — CHANGES_REQUESTED** [review]
ok, but mark is too low. I remember you were one of those who wanted to make recusrsive crawler for 10. keep going!

File: lab_5_scrapper/scrapper.py (hunk 1 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper.py]: you should not put it in instance's attribute. nobody stops from making the instance inside the method and re-use only for this method.

File: lab_5_scrapper/scrapper.py (hunk 9 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:331]: fixed

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:328]: fixed

File: lab_5_scrapper/scrapper.py (hunk 12 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:250]: fixed (if it was ment to do response.ok)

File: lab_5_scrapper/scrapper.py (hunk 1 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py]: done

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:98]: done

File: requirements.txt (hunk 4 above)
Comment (EkaterinaKhomutova) [reply requirements.txt:2]: ??

File: lab_5_scrapper/settings.json (hunk 6 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/settings.json:2]: I will make it higher

File: lab_5_scrapper/scrapper_config.json (hunk 7 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper_config.json:2]: fixed

File: lab_5_scrapper/scrapper_config.json (hunk 8 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper_config.json:3]: done

File: lab_5_scrapper/scrapper.py (hunk 14 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:251]: fixed

File: lab_5_scrapper/scrapper.py (hunk 15 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:235]: I haven't found a better strategy, on my web site there no classes in a tags and there is no unique class for news container

File: lab_5_scrapper/scrapper.py (hunk 16 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:240]: fixed

File: lab_5_scrapper/scrapper.py (hunk 17 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:88]: done

File: lab_5_scrapper/scrapper.py (hunk 13 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:252]: fixed

File: requirements.txt (hunk 4 above)
Comment (demid5111) [reply requirements.txt:2]: do you think this is a proper package that you need to install?

File: lab_5_scrapper/scrapper.py [hunk 18]
```diff
@@ -204,12 +382,29 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if not base_path.exists():
+        base_path.mkdir(parents=True)
+    else:
+        shutil.rmtree(base_path)
+        base_path.mkdir(parents=True, exist_ok=True)
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:389]: it is enough just to remove dir and then create it from scratch

File: lab_5_scrapper/scrapper.py (hunk 10 above)
Comment (demid5111) [reply lab_5_scrapper/scrapper.py]: no

File: lab_5_scrapper/scrapper.py [hunk 19]
```diff
@@ -127,11 +234,39 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
...
         Find articles.
         """
+        seed_urls = self.get_search_urls()
+        necessary_len = self.config.get_num_articles()
+
+        while len(self.urls) != necessary_len:
+            for url in seed_urls:
+                response = make_request(url, self.config)
+
+                if response.ok:
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:255]: rework to have code less nested (skip iteration if not OK)

File: lab_5_scrapper/scrapper.py [hunk 20]
```diff
@@ -127,11 +234,39 @@ def _extract_url(self, article_bs: BeautifulSoup) -> str:
         Returns:
             str: Url from HTML
         """
+        container_div = article_bs.find('div', class_='col-lg-9 col-md-12')
+        for link in container_div.find_all('a', href=True):
+            if link.get('href').startswith('/news/202'):
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:239]: less nested

File: lab_5_scrapper/scrapper.py [hunk 21]
```diff
@@ -176,6 +330,23 @@ def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> No
         Args:
             article_soup (bs4.BeautifulSoup): BeautifulSoup instance
         """
+        header = article_soup.find('h1')
+        if header:
+            self.article.title = header
+
+        self.article.author = 'NOT FOUND'
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:337]: author should be a list? what is written in docs

File: lab_5_scrapper/scrapper.py [hunk 22]
```diff
@@ -176,6 +330,23 @@ def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> No
...
+
+        element = article_soup.find('ul', class_='list-unstyled list-inline').find('li', class_='list-inline-item')
+        date = element.get_text().strip().split(', ')[1]
+        self.article.date = date
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:341]: date is needed for 8, do you plan for 8?

File: lab_5_scrapper/scrapper.py [hunk 23]
```diff
@@ -176,6 +330,23 @@ def _fill_article_with_meta_information(self, article_soup: BeautifulSoup) -> No
...
+
+        meta_tag = article_soup.find('meta', attrs={'meta': 'keywords'})
+        topics = []
```
Comment (demid5111) [inline lab_5_scrapper/scrapper.py:344]: topic is needed for 8, do you plan for 8?

**Review by demid5111 — CHANGES_REQUESTED** [review]
you almost made everything for 8. maybe increasing the mark in config?

File: lab_5_scrapper/scrapper.py (hunk 18 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:389]: done

File: lab_5_scrapper/scrapper.py (hunk 19 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:255]: done

File: lab_5_scrapper/scrapper.py (hunk 20 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:239]: done

File: lab_5_scrapper/scrapper.py (hunk 21 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:337]: yes, it should be, I fixed

File: lab_5_scrapper/scrapper.py (hunk 22 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:341]: yes

File: lab_5_scrapper/scrapper.py (hunk 23 above)
Comment (EkaterinaKhomutova) [reply lab_5_scrapper/scrapper.py:344]: yes

**Review by demid5111 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**